### PR TITLE
Add Arun metric

### PR DESCRIPTION
**Pull request by souravsingh**
Fixes a part of #1275

### CODE REVIEW

**Comment by souravsingh** [comment]
@menshikh-iv @piskvorky Can you take a look at the PR and suggest any improvements?

File: gensim/matutils.py [hunk 1]
```diff
@@ -479,6 +479,38 @@ def isbow(vec):
         return False
     return True
 
+def symmetric_kl(distrib_p, distrib_q):
+    return numpy.sum([stats.entropy(distrib_p, distrib_q), stats.entropy(distrib_p, distrib_q)])
```
Comment (piskvorky) [inline gensim/matutils.py]: Why `numpy.sum` instead of `+`?

File: gensim/matutils.py [hunk 2]
```diff
@@ -479,6 +479,38 @@ def isbow(vec):
...
+    
+    :param min_num_topics: Minimum number of topics to test
+    :param max_num_topics: Maximum number of topics to test
+    :param iterations: Number of iterations per value of k
+    :return: A list of len (max_num_topics - min_num_topics) with the average symmetric KL divergence for each k
+    
+    Thanks to Adrien Guille for the implementation of the metric.
+    """
+    kl_matrix = []
+    for j in range(iterations):
```
Comment (piskvorky) [inline gensim/matutils.py]: `xrange` more efficient.

File: gensim/matutils.py [hunk 3]
```diff
@@ -479,6 +479,38 @@ def isbow(vec):
...
+        kl_list = []
+        l = np.array([sum(self.corpus.vector_for_document(doc_id)) for doc_id in range(self.corpus.size)])  # document length
```
Comment (piskvorky) [inline gensim/matutils.py]: `xrange` more efficient.

Also, the whole construct is too complicated -- what does this do? The comment is not very helpful.

File: gensim/matutils.py [hunk 4]
```diff
@@ -479,6 +479,38 @@ def isbow(vec):
         return False
     return True
 
+def symmetric_kl(distrib_p, distrib_q):
+    return numpy.sum([stats.entropy(distrib_p, distrib_q), stats.entropy(distrib_p, distrib_q)])
+
+def arun_metric(self, min_num_topics=10, max_num_topics=50, iterations=10):
```
Comment (piskvorky) [inline gensim/matutils.py]: What is `self`? This is not an object method.

File: gensim/matutils.py [hunk 5]
```diff
@@ -479,6 +479,38 @@ def isbow(vec):
...
+    """
+    Implements Arun metric to estimate the optimal number of topics:
+    Arun, R., V. Suresh, C. V. Madhavan, and M. N. Murthy
+    On finding the natural number of topics with latent dirichlet allocation: Some observations.
+    In PAKDD (2010), pp. 391–402.
+    
+    :param min_num_topics: Minimum number of topics to test
```
Comment (piskvorky) [inline gensim/matutils.py]: Google-style docstrings please.

File: gensim/matutils.py [hunk 6]
```diff
@@ -479,6 +479,38 @@ def isbow(vec):
...
+    :param max_num_topics: Maximum number of topics to test
+    :param iterations: Number of iterations per value of k
+    :return: A list of len (max_num_topics - min_num_topics) with the average symmetric KL divergence for each k
```
Comment (piskvorky) [inline gensim/matutils.py]: Does `length of (max - min)` mean one of range boundaries is exclusive? Which one?

File: gensim/matutils.py [hunk 7]
```diff
@@ -479,6 +479,38 @@ def isbow(vec):
...
+        norm = np.linalg.norm(l)
+        for i in range(min_num_topics, max_num_topics + 1):
+            self.infer_topics(i)
+            c_m1 = np.linalg.svd(self.topic_word_matrix.todense(), compute_uv=False)
+            c_m2 = l.dot(self.document_topic_matrix.todense())
+            c_m2 += 0.0001  
+            c_m2 /= norm
+            kl_list.append(symmetric_kl(c_m1.tolist(), c_m2.tolist()[0]))
+        kl_matrix.append(kl_list)
+    output = np.array(kl_matrix)
```
Comment (piskvorky) [inline gensim/matutils.py]: Docstring says the result of this function is a list of average divergences (I'd assume numbers, floats). But this line implies some array of matrices? What does this actually do?

**Comment by souravsingh** [comment]
@piskvorky The metric is used for finding the optimal minimum and maximum number of topics for LDA models. I have an example for the Arun metric in R, which I can show.

File: gensim/matutils.py (hunk 6 above)
Comment (souravsingh) [inline gensim/matutils.py]: It is exclusive for of ``min_num_topics``

File: gensim/matutils.py (hunk 1 above)
Comment (souravsingh) [inline gensim/matutils.py]: I was thinking that ``numpy.sum`` could be faster than ``+`` for the NumPy matrix operation.

File: gensim/matutils.py [hunk 8]
```diff
@@ -479,6 +479,42 @@ def isbow(vec):
...
+    Returns: 
+        A list of len (max_num_topics - min_num_topics) with the average symmetric KL divergence for each k
+    
+    Thanks to Adrien Guille for the implementation of the metric.
+    
+    """
+    kl_matrix = []
+    for j in xrange(iterations):
+        kl_list = []
+        l = np.array([sum(corpus.vector_for_document(doc_id)) for doc_id in xrange(corpus.size)])  # document length
```
Comment (menshikh-iv) [inline gensim/matutils.py]: `corpus` is undefined

File: gensim/matutils.py [hunk 9]
```diff
@@ -479,6 +479,42 @@ def isbow(vec):
...
+        norm = np.linalg.norm(l)
+        for i in xrange(min_num_topics, max_num_topics + 1):
+            c_m1 = np.linalg.svd(self.topic_word_matrix.todense(), compute_uv=False)
```
Comment (menshikh-iv) [inline gensim/matutils.py]: `self` is undefined, this function isn't a method of any class -> `topic_word_matrix` is undefined too.

File: gensim/matutils.py [hunk 10]
```diff
@@ -479,6 +479,42 @@ def isbow(vec):
         return False
     return True
 
+def symmetric_kl(distrib_p, distrib_q):
+    return numpy.sum([stats.entropy(distrib_p, distrib_q), stats.entropy(distrib_p, distrib_q)])
+
+def arun_metric(min_num_topics=10, max_num_topics=50, iterations=10):
```
Comment (menshikh-iv) [inline gensim/matutils.py]: Looks like you missed several arguments.

**Comment by menshikh-iv** [comment]
This code doesn't work now, please fix it.
Also, don't forget to add tests and example in lda notebook.

File: gensim/matutils.py (hunk 1 above)
Comment (piskvorky) [reply gensim/matutils.py]: How is it a matrix operation? It seems to add two numbers.

Comment (souravsingh) [reply gensim/matutils.py]: The ``distrib_p`` and ``distrib_q`` are two matrices from which the Symmetric KL divergence is calculated. The exact formula(which is adapted from the R package) is ``divergence <- sum(cm1*log(cm1/cm2)) + sum(cm2*log(cm2/cm1))`` which is similar to what this function does.

Comment (piskvorky) [reply gensim/matutils.py]: That's a non-standard use of KL (which works over a probability distribution = vector). 
Deserves a comment & proper docstring.

What is actually `distrib_p` and `distrib_q`? Why are they matrices?

**Comment by menshikh-iv** [comment]
This code is copy-paste from [AdrienGuille/TOM](https://github.com/AdrienGuille/TOM/blob/dbff857aa50d1f337909ad2091612e134b78ecde/tom_lib/nlp/topic_model.py#L63)

**Comment by souravsingh** [comment]
@menshikh-iv I have obtained permission for code reuse from the dev and have added his name as an acknowledgement in the docstring.

**Comment by souravsingh** [comment]
@menshikh-iv The code for Arun Metric accepts a DTM(Document-Term Matrix) to know the optimal number of topics, How do we take the DTM matrix for the Arun metric?

**Comment by menshikh-iv** [comment]
@souravsingh after `Dictionary.doc2bow` you receive bag-of-word representation, it's a part of document-term matrices.

Fit `Dictionary`, use `doc2bow` and convert this representation to matrix, that's enough.

**Comment by menshikh-iv** [comment]
What's a status here @souravsingh ?

**Comment by souravsingh** [comment]
@menshikh-iv I have a script ready which can calculate Arun metric from the model. I will have an output ready(which will be a graph) in a few hours.

File: gensim/matutils.py [hunk 11]
```diff
@@ -485,9 +485,9 @@ def symmetric_kl(distrib_p, distrib_q, num_features=None):
     Used to measure the divergence between two probability distributions
     """
     distrib_p, distrib_q = convert_vec(distrib_p, distrib_q, num_features= num_features)
-    return (stats.entropy(distrib_p, distrib_q) + stats.entropy(distrib_q, distrib_p))
+    return (numpy.sum([stats.entropy(p, q), stats.entropy(q, p)]))
```
Comment (menshikh-iv) [inline gensim/matutils.py]: `p`, `q` is undefined here

File: gensim/matutils.py [hunk 12]
```diff
@@ -505,21 +505,22 @@ def arun_metric(min_num_topics=10, max_num_topics=50, iterations=10):
...
-            c_m2 += 0.0001  # we need this to prevent components equal to zero
-            c_m2 /= norm
-            kl_list.append(symmetric_kl(c_m1.tolist(), c_m2.tolist()[0]))
-        kl_matrix.append(kl_list)
-    output = np.array(kl_matrix)
-    return output.mean(axis=0)
-
+    results = []
+    for i in range(min_topics, max_topics, iteration):
+        lda = models.ldamodel.LdaModel(    # Create an LDA model instance
```
Comment (menshikh-iv) [inline gensim/matutils.py]: `LdaMulticore` is faster

File: gensim/matutils.py [hunk 13]
```diff
@@ -505,21 +505,22 @@ def arun_metric(min_num_topics=10, max_num_topics=50, iterations=10):
...
+            corpus=corpus,
+            id2word=dictionary,
+            num_topics=i
+        )
+    U, document_word_vector, V = np.linalg.svd(lda.expElogbeta)
+    lda_topic = lda[corpus] # Get topics
+    term_document_matrix = matutils.corpus2dense(lda_topic, lda.num_topic).transpose() # Create DTM matrix
+    corpus_length_vector = np.array([sum(frequency for _, frequency in document) for document in my_corpus])
+    document_topic_vector = corpus_length_vector.dot(term_document_matrix)
+    document_topic_norm   = np.linalg.norm(corpus_length_vector)
```
Comment (menshikh-iv) [inline gensim/matutils.py]: only one space needed before ` =`

File: gensim/matutils.py [hunk 14]
```diff
@@ -505,21 +505,22 @@ def arun_metric(min_num_topics=10, max_num_topics=50, iterations=10):
...
-            c_m2 = l.dot(self.document_topic_matrix.todense())
-            c_m2 += 0.0001  # we need this to prevent components equal to zero
-            c_m2 /= norm
-            kl_list.append(symmetric_kl(c_m1.tolist(), c_m2.tolist()[0]))
-        kl_matrix.append(kl_list)
-    output = np.array(kl_matrix)
-    return output.mean(axis=0)
-
+    results = []
+    for i in range(min_topics, max_topics, iteration):
```
Comment (menshikh-iv) [inline gensim/matutils.py]: For big models, it needs a lot of time. It doesn't look like very useful feature for real tasks.

File: gensim/matutils.py [hunk 15]
```diff
@@ -505,21 +505,22 @@ def arun_metric(min_num_topics=10, max_num_topics=50, iterations=10):
...
+            num_topics=i
+        )
+    U, document_word_vector, V = np.linalg.svd(lda.expElogbeta)
+    lda_topic = lda[corpus] # Get topics
+    term_document_matrix = matutils.corpus2dense(lda_topic, lda.num_topic).transpose() # Create DTM matrix
+    corpus_length_vector = np.array([sum(frequency for _, frequency in document) for document in my_corpus])
+    document_topic_vector = corpus_length_vector.dot(term_document_matrix)
+    document_topic_norm   = np.linalg.norm(corpus_length_vector)
+    document_topic_vector = document_topic_vector / document_topic_norm
+    result.append(symmetric_kl_divergence(document_word_vector, document_topic_vector))
```
Comment (menshikh-iv) [inline gensim/matutils.py]: What means result list? it's not mapping `num_topics` to value, it's list with values.

**Comment by souravsingh** [comment]
@menshikh-iv I am working on finishing up the implementation of the metric, but here is the plot I obtain.
kl_topics

The script used for obtaining the plot is here- https://gist.github.com/souravsingh/bd7950aa07a53b651aa1fcc4b5b2e5e8

File: gensim/matutils.py (hunk 15 above)
Comment (souravsingh) [reply gensim/matutils.py]: The list will be the output of the function.

File: gensim/matutils.py (hunk 14 above)
Comment (souravsingh) [reply gensim/matutils.py]: Agree on this. I thought of taking the trained model as a parameter, but that makes it difficult to evaluate the range

**Comment by souravsingh** [comment]
@menshikh-iv I have made a few fixes. I would need some help in finishing this up.

**Comment by menshikh-iv** [comment]
@souravsingh your plot is empty, please fix it.
I think your code will be good as part of the concrete notebook, for example [lda training tips](https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/lda_training_tips.ipynb).

**Comment by menshikh-iv** [comment]
@souravsingh please create new PR and add this technique to the [lda training tips](https://github.com/RaRe-Technologies/gensim/blob/develop/docs/notebooks/lda_training_tips.ipynb) notebook (with a description and examples).

**Pull request closed without merging**